### PR TITLE
[GPU] Enable fc 4d for MatMul

### PR DESCRIPTION
**Pull request by steve-y**
### Details:
 - *Enable fc 4d for MatMul to calculate 4x2*

### Tickets:
 - *132334*

### CODE REVIEW

File: src/plugins/intel_gpu/tests/unit/shape_infer/matmul_si_test.cpp [hunk 1]
```diff
@@ -120,6 +120,24 @@ INSTANTIATE_TEST_SUITE_P(smoke, fully_connected_test,
...
+            layout{ov::PartialShape{10, 10, 10, 1024}, data_types::f32, format::bfyx},
+            layout{ov::PartialShape{1000, 1024}, data_types::f32, format::bfyx},
+            data_types::f32, false, false,
+            layout{ov::PartialShape{10, 10, 10, 1000}, data_types::f32, format::bfyx}
+        },
+        {
+            layout{ov::PartialShape::dynamic(4), data_types::f32, format::bfyx},
+            layout{ov::PartialShape::dynamic(2), data_types::f32, format::bfyx},
+            data_types::f32, false, false,
+            layout{ov::PartialShape::dynamic(4), data_types::f32, format::bfyx}
```
Comment (andrew-k-park) [inline src/plugins/intel_gpu/tests/unit/shape_infer/matmul_si_test.cpp:140]: These test params do not cover your changes in `calc_output_layout`. Please add function test case with 4d input + 2d weight shapes for checking shape inference and accuracy together.

File: src/plugins/intel_gpu/src/graph/fully_connected.cpp [hunk 2]
```diff
@@ -117,7 +117,7 @@ layout fully_connected_inst::calc_output_layout(fully_connected_node const& node
         feature = std::max({input_layout.spatial(0), input_layout.spatial(1), input_layout.spatial(2)});
     }
 
-    if (desc->input_size > 3) {
+    if (desc->input_size > 4) {
        input_layout.set_partial_shape(reshape_to_2d(input_pshape, feature));
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/graph/fully_connected.cpp:121]: If shape is 5d, e.g.,
In [1,3,5,10,4] W [1,3,5,4,20]
feature is detected as inputshape[3] => 10. 
However actually it is expected to be recognized as 4. 
Seems this is not still covered by the test.

Comment (steve-y) [reply src/plugins/intel_gpu/src/graph/fully_connected.cpp:121]: Yes, this PR was extended fc 3d case to support fc 4d case.

File: src/plugins/intel_gpu/tests/functional/single_layer_tests/dynamic/matmul.cpp [hunk 3]
```diff
@@ -335,6 +335,18 @@ const auto fullyConnectedParams3D_nightly = ::testing::Combine(::testing::Values
...
+    {ov::test::static_shapes_to_test_representation({{1, 32, 20, 120}, {120, 3}}), {true, true}},
+
+    {ov::test::static_shapes_to_test_representation({{1, 1, 4, 50}, {50, 7}}), {false, false}},
+    {ov::test::static_shapes_to_test_representation({{1, 1, 4, 50}, {50, 7}}), {true, true}},
+
+    {ov::test::static_shapes_to_test_representation({{1, 1, 5, 32}, {32, 3}}), {false, false}},
+    {ov::test::static_shapes_to_test_representation({{1, 1, 5, 32}, {32, 3}}), {true, false}},
+
+    {ov::test::static_shapes_to_test_representation({{1, 3, 6, 14}, {14, 10}}), {true, false}},
+    {ov::test::static_shapes_to_test_representation({{1, 3, 6, 14}, {14, 10}}), {false, true}},
```
Comment (yeonbok) [inline src/plugins/intel_gpu/tests/functional/single_layer_tests/dynamic/matmul.cpp:348]: None of these tests are failing after reverting the above changes. Please add a meaningful test which fails beofre your change, but fixed after your change.

Comment (steve-y) [reply src/plugins/intel_gpu/tests/functional/single_layer_tests/dynamic/matmul.cpp]: A unit test for this fix was checked and added, and the previous functional tests were removed as they were not proper as you commented. Thanks for your comment.

**Review by yeonbok — APPROVED** [review]

**Pull request merged**